homeassistant/components/binary_sensor/zwave.py

```diff
@@ -18,31 +18,30 @@
 DEPENDENCIES = []
 
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
-    """Setup the Z-Wave platform for binary sensors."""
-    if discovery_info is None or zwave.NETWORK is None:
-        return
-
-    node = zwave.NETWORK.nodes[discovery_info[zwave.const.ATTR_NODE_ID]]
-    value = node.values[discovery_info[zwave.const.ATTR_VALUE_ID]]
+def get_device(value, hass, **kwargs):
+    """Create zwave entity device."""
     value.set_change_verified(False)
 
     device_mapping = workaround.get_device_mapping(value)
     if device_mapping == workaround.WORKAROUND_NO_OFF_EVENT:
         # Default the multiplier to 4
         re_arm_multiplier = (zwave.get_config_value(value.node, 9) or 4)
-        add_devices([
-            ZWaveTriggerSensor(value, "motion",
-                               hass, re_arm_multiplier * 8)
-        ])
-        return
+        return ZWaveTriggerSensor(value, "motion", hass, re_arm_multiplier * 8)
 
     if workaround.get_device_component_mapping(value) == DOMAIN:
-        add_devices([ZWaveBinarySensor(value, None)])
-        return
+        return ZWaveBinarySensor(value, None)
 
     if value.command_class == zwave.const.COMMAND_CLASS_SENSOR_BINARY:
-        add_devices([ZWaveBinarySensor(value, None)])
+        return ZWaveBinarySensor(value, None)
+    return None
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```

Comment:
I think that you can make a generic setup that you just import for each platform. We should make it async for improved speed:

(to be placed in `zwave/__init__.py`)

```python
@asyncio.coroutine
def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
    """Generic Z-Wave platform setup."""
    if discovery_info is None or zwave.NETWORK is None:
        return False
    device = zwave.get_device(hass, discovery_info[zwave.const.DISCOVERY_DEVICE])
    if device:
        yield from async_add_devices([device])
        return True
    else:
        return False
```

In the case of the lock component, you could then wrap this method and register the services.

```python
from homeassistant.components.zwave import async_setup_platform as zwave_async_setup_platform
@asyncio.coroutine
def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
    result = yield from zwave_async_setup_platform(…)
    if not result or hass.services.has_service('lock', SERVICE_XX):
        return
    # Register services
```